File: event_hub_spring/src/main/java/com/eventhub/entity/Booking.java
package com.eventhub.entity;

import jakarta.persistence.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@Table(name = "bookings")
@EntityListeners(AuditingEntityListener.class)
public class Booking {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "booking_id", updatable = false, nullable = false)
    private String bookingId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "event_id", nullable = false)
    private Event event;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "ticket_id", nullable = false)
    private Ticket ticket;

    @Column(nullable = false)
    private Integer quantite;

    @Column(name = "montant_total", nullable = false)
    private Double montantTotal;

    @Column(nullable = false)
    private String devise = "USD";

    @Column(nullable = false)
    private String statut = "pending";

    @Column(name = "qr_code_token")
    private String qrCodeToken;

    @Column(name = "qr_code_url")
    private String qrCodeURL;

    @Column(name = "pdf_url")
    private String pdfURL;

    @Column(name = "payment_id")
    private String paymentId;

    @Column(name = "scanned_at")
    private LocalDateTime scannedAt;

    @CreatedDate
    @Column(name = "date_reservation", nullable = false, updatable = false)
    private LocalDateTime dateReservation;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public Booking() {}

    public String getBookingId() { return bookingId; }
    public void setBookingId(String bookingId) { this.bookingId = bookingId; }
    public User getUser() { return user; }
    public void setUser(User user) { this.user = user; }
    public Event getEvent() { return event; }
    public void setEvent(Event event) { this.event = event; }
    public Ticket getTicket() { return ticket; }
    public void setTicket(Ticket ticket) { this.ticket = ticket; }
    public Integer getQuantite() { return quantite; }
    public void setQuantite(Integer quantite) { this.quantite = quantite; }
    public Double getMontantTotal() { return montantTotal; }
    public void setMontantTotal(Double montantTotal) { this.montantTotal = montantTotal; }
    public String getDevise() { return devise; }
    public void setDevise(String devise) { this.devise = devise; }
    public String getStatut() { return statut; }
    public void setStatut(String statut) { this.statut = statut; }
    public String getQrCodeToken() { return qrCodeToken; }
    public void setQrCodeToken(String qrCodeToken) { this.qrCodeToken = qrCodeToken; }
    public String getQrCodeURL() { return qrCodeURL; }
    public void setQrCodeURL(String qrCodeURL) { this.qrCodeURL = qrCodeURL; }
    public String getPdfURL() { return pdfURL; }
    public void setPdfURL(String pdfURL) { this.pdfURL = pdfURL; }
    public String getPaymentId() { return paymentId; }
    public void setPaymentId(String paymentId) { this.paymentId = paymentId; }
    public LocalDateTime getScannedAt() { return scannedAt; }
    public void setScannedAt(LocalDateTime scannedAt) { this.scannedAt = scannedAt; }
    public LocalDateTime getDateReservation() { return dateReservation; }
    public void setDateReservation(LocalDateTime dateReservation) { this.dateReservation = dateReservation; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }

    public boolean isPending() { return "pending".equals(statut); }
    public boolean isConfirmed() { return "confirmed".equals(statut); }
    public boolean isCancelled() { return "cancelled".equals(statut); }
    public boolean isRefunded() { return "refunded".equals(statut); }
    public boolean isUsed() { return "used".equals(statut); }
    public boolean isScanned() { return scannedAt != null; }
}

File: event_hub_back/lib/models/booking.dart
import 'event.dart';
import 'ticket.dart';

class Booking {
  final String bookingId;
  final String userId;
  final String eventId;
  final String ticketId;
  final int quantite;
  final double montantTotal;
  final String devise;
  final String statut;
  final String? qrCodeToken;
  final String? qrCodeURL;
  final String? pdfURL;
  final String? paymentId;
  final DateTime? scannedAt;
  final DateTime dateReservation;
  final DateTime updatedAt;

  // Populated fields
  final Event? event;
  final Ticket? ticket;

  Booking({
    required this.bookingId,
    required this.userId,
    required this.eventId,
    required this.ticketId,
    required this.quantite,
    required this.montantTotal,
    this.devise = 'USD',
    this.statut = 'pending',
    this.qrCodeToken,
    this.qrCodeURL,
    this.pdfURL,
    this.paymentId,
    this.scannedAt,
    required this.dateReservation,
    required this.updatedAt,
    this.event,
    this.ticket,
  });

  bool get isPending => statut == 'pending';
  bool get isConfirmed => statut == 'confirmed';
  bool get isCancelled => statut == 'cancelled';
  bool get isRefunded => statut == 'refunded';
  bool get isUsed => statut == 'used';
  bool get isScanned => scannedAt != null;

  factory Booking.fromMap(Map<String, dynamic> map) {
    Event? event;
    Ticket? ticket;

    if (map['event'] != null && map['event'] is Map) {
      event = Event.fromMap(Map<String, dynamic>.from(map['event']));
    }

    if (map['ticket'] != null && map['ticket'] is Map) {
      ticket = Ticket.fromMap(Map<String, dynamic>.from(map['ticket']));
    }

    return Booking(
      bookingId: map['bookingId'] ?? map['id'] ?? '',
      userId: map['userId'] ?? '',
      eventId: map['eventId'] ?? '',
      ticketId: map['ticketId'] ?? '',
      quantite: map['quantite'] ?? 0,
      montantTotal: (map['montantTotal'] ?? 0).toDouble(),
      devise: map['devise'] ?? 'USD',
      statut: map['statut'] ?? 'pending',
      qrCodeToken: map['qrCodeToken'],
      qrCodeURL: map['qrCodeURL'],
      pdfURL: map['pdfURL'],
      paymentId: map['paymentId'],
      scannedAt: map['scannedAt']?.toDate(),
      dateReservation: map['dateReservation']?.toDate() ?? DateTime.now(),
      updatedAt: map['updatedAt']?.toDate() ?? DateTime.now(),
      event: event,
      ticket: ticket,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'bookingId': bookingId,
      'userId': userId,
      'eventId': eventId,
      'ticketId': ticketId,
      'quantite': quantite,
      'montantTotal': montantTotal,
      'devise': devise,
      'statut': statut,
      'qrCodeToken': qrCodeToken,
      'qrCodeURL': qrCodeURL,
      'pdfURL': pdfURL,
      'paymentId': paymentId,
      'scannedAt': scannedAt,
      'dateReservation': dateReservation,
      'updatedAt': updatedAt,
    };
  }
}

class BookingWithClientSecret {
  final String bookingId;
  final String? clientSecret;
  final String? paymentId;
  final double montantTotal;
  final String statut;

  BookingWithClientSecret({
    required this.bookingId,
    this.clientSecret,
    this.paymentId,
    required this.montantTotal,
    required this.statut,
  });

  factory BookingWithClientSecret.fromMap(Map<String, dynamic> map) {
    return BookingWithClientSecret(
      bookingId: map['bookingId'] ?? '',
      clientSecret: map['clientSecret'],
      paymentId: map['paymentId'],
      montantTotal: (map['montantTotal'] ?? 0).toDouble(),
      statut: map['statut'] ?? 'pending',
    );
  }
}
